src/shogun/io/ARFFFile.h

```diff
@@ -0,0 +1,545 @@
+		{
+			if (java_token == ':')
+				return ":";
+			if (java_token == '\'')
+				return "";
+			if (java_token == '-')
+				return "-";
+			if (java_token == ' ')
+				return " ";
+			return nullptr;
+		}
+
+		/**
+		 * Checks if a Java token is valid and returns string representing the
+		 * C++ token.
+		 * @param java_time_token token to check and translate
+		 * @return translated C++ token
+		 */
+		SG_FORCED_INLINE const char*
+		check_and_append_j2cpp(const std::string& java_time_token)
+		{
+			if (auto cpp_token = process_javatoken(java_time_token))
+				return cpp_token;
+			else
+				SG_SERROR(
+				    "Could not convert Java time token \"%s\" to C++ time "
+				    "token.\n",
+				    java_time_token.c_str())
+			return nullptr;
+		}
+
+		SG_FORCED_INLINE const char*
+		check_and_append_j2cpp(char java_time_token)
+		{
+			if (auto cpp_token = process_javatoken(java_time_token))
+				return cpp_token;
+			else
+				SG_SERROR(
+				    "Could not convert Java time token \"%c\" to C++ time "
+				    "token.\n",
+				    java_time_token)
+			return nullptr;
+		}
+
+		/**
+		 * Converts a Java SimpleDateFormat to a C++ date format
+		 * @param java_time the string to translate
+		 * @return the C++ format equivalent of java_time
+		 */
+		SG_FORCED_INLINE std::string
+		javatime_to_cpptime(const std::string& java_time)
+		{
+			std::string cpp_time;
+			std::string token;
+			auto begin = java_time.begin();
+			auto it = java_time.begin();
+			while (it != java_time.end())
+			{
+				if (*it == '-' || *it == ' ' || *it == ':')
+				{
+					token = {begin, it};
+					cpp_time.append(check_and_append_j2cpp(token));
+					cpp_time.append(check_and_append_j2cpp(*it));
+					begin = std::next(it);
+				}
+				else if (*it == '\'')
+				{
+					token = {begin, it};
+					cpp_time.append(check_and_append_j2cpp(token));
+					cpp_time.append(check_and_append_j2cpp(*it));
+					begin = std::next(it);
+					begin = it;
+					it = std::next(it);
+					while (*it != '\'')
+					{
+						it = std::next(it);
+					}
+					token = {std::next(begin), it};
+					cpp_time.append(token);
+					begin = std::next(it);
+				}
+				else if (std::next(it) == java_time.end())
+				{
+					token = {begin, std::next(it)};
+					if (auto cpp_token = process_javatoken(token))
+					{
+						cpp_time.append(cpp_token);
+					}
+					else
+						SG_SERROR(
+						    "Could not convert Java time token %s to C++ time "
+						    "token.\n",
+						    token.c_str())
+				}
+				it = std::next(it);
+			}
+			return cpp_time;
+		}
+	} // namespace arff_detail
+	/**
+	 * ARFFDeserializer parses files in the ARFF format.
+	 * For information about this format see
+	 * https://waikato.github.io/weka-wiki/arff_stable/
+	 */
+	class ARFFDeserializer
+	{
+	private:
+		/**
+		 * The attributes supported in the ARFF format
+		 */
+		enum class Attribute
+		{
+			Numeric = 0,
+			Integer = 1,
+			Real = 2,
+			String = 3,
+			Date = 4,
+			Nominal = 5
```

Comment:
I guess we should add these type of feautres to shogun ... would be another argument to use it over some of the other libs that need everything in real matrix form